[Hub Generated] Review request for Microsoft.ResourceHealth to add version preview/2023-07-01-preview

mkatkam opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/mkatkam-resourcehealth-Microsoft.ResourceHealth-2023-07-01-preview...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5e91ff2d-1f5c-4e96-87c2-390e7451da37)

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I have reviewed the general guidance on the spec PR review process: https://aka.ms/specprreview.
- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data-plane related specifications.
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy).
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can request review from an ARM API Review board.

### ARM API changes review

- If you want for the ARM team to review this PR, you must add the `ARMReview` label. 
- The automation may automatically add the `ARMReview` label, if appropriate.  
  If this happens, proceed according to guidance given in GitHub comments also added by the automation.

### Breaking change review

If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), 
follow the process outlined in the [High-level Breaking Change Process doc](https://aka.ms/breakingchangesprocess#high-level-breaking-change-process).
      
## Getting help

- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For additional help, see https://aka.ms/azsdk/support/spectools.

mkatkam committed:
Adds base for updating Microsoft.ResourceHealth from version preview/2022-10-01-preview to version 2023-07-01-preview

mkatkam committed:
Updates readme

mkatkam committed:
Updates API version in new specs and examples

mkatkam committed:
Adding 3 new fields maintenanceId, maintenanceType and argQuery to Events Object and modified examples accordingly

abhishek-corp commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/examples/Events_ListBySingleResource.json:
```diff
@@ -93,7 +93,10 @@
               "priority": 2,
               "lastUpdateTime": "2018-11-08T00:00:00Z",
-              "hirStage": "active"
+              "hirStage": "active",
```
We don't need to show these new fields in this API example. The API flow is different for SingleResource events.

abhishek-corp commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -1197,6 +1197,18 @@
+            },
+            "maintenanceId": {
+              "type": "string",
+              "description": "Maintenance Id"
```
Put better description so external users can understand the use case.
Something like
"This is a unique id for Planned Maintenance events used by some of our systems. "

abhishek-corp commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -1197,6 +1197,18 @@
+              "type": "string",
+              "description": "Maintenance Id"
+            },
+            "maintenanceType": {
```
Put better description so external users can understand the use case.

abhishek-corp commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -1197,6 +1197,18 @@
+            },
+            "argQuery": {
+              "type": "string",
+              "description": "ARG Query"
```
Put better description so external users can understand the use case.

abhishek-corp commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+            },
+            "duration": {
+              "type": "integer",
+              "description": "duration in seconds",
```
Can we change this to something like this?

"Approx duration of impact in seconds. Mostly used for Planned Maintenance events."

abhishek-corp commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+            },
+            "impactType": {
+              "type": "string",
+              "description": "The type of the impact"
```
Can we change this to something like this?

"The type of impact. Mostly used for Planned Maintenance events."

mkatkam replied:
Keeping this as is for now, to avoid any  breaking changes.

visingla-ms commented:
> Hi @mkatkam! The automation detected breaking changes in this pull request. As a result, it added the `BreakingChangeReviewRequired` label. You cannot proceed with merging this PR until you complete one of the following action items: **ACTION ITEM ALTERNATIVE A**: Fix the breaking change. Please consult the documentation provided in the relevant validation failures. **ACTION ITEM ALTERNATIVE B**: Request approval. Alternatively, if you cannot fix the breaking changes, then you can request an approval for them. Please follow the process described in the [High-level Breaking Change Process doc](https://aka.ms/breakingchangesprocess#high-level-breaking-change-process). **ACTION ITEM ALTERNATIVE C**: Report false positive. If you think there are no breaking changes, i.e. the validation should pass yet it fails, then proceed as explained in **ACTION ITEM ALTERNATIVE B**. This applies even if the breaking change tool fails with internal runtime error. In such case a manual breaking change review is necessary.

Please fix the breaking changes or please get a breaking changes approval from Azure Breaking Changes Reviewers <azbreakchangereview@microsoft.com> and ARM can review the changes afterwards. Once you have the approval, please remove the "ARMChangesRequested" label from the PR to make the PR visible to the ARM reviewers.

mkatkam commented:
> > Hi @mkatkam! The automation detected breaking changes in this pull request. As a result, it added the `BreakingChangeReviewRequired` label. You cannot proceed with merging this PR until you complete one of the following action items: **ACTION ITEM ALTERNATIVE A**: Fix the breaking change. Please consult the documentation provided in the relevant validation failures. **ACTION ITEM ALTERNATIVE B**: Request approval. Alternatively, if you cannot fix the breaking changes, then you can request an approval for them. Please follow the process described in the [High-level Breaking Change Process doc](https://aka.ms/breakingchangesprocess#high-level-breaking-change-process). **ACTION ITEM ALTERNATIVE C**: Report false positive. If you think there are no breaking changes, i.e. the validation should pass yet it fails, then proceed as explained in **ACTION ITEM ALTERNATIVE B**. This applies even if the breaking change tool fails with internal runtime error. In such case a manual breaking change review is necessary.
> 
> Please fix the breaking changes or please get a breaking changes approval from Azure Breaking Changes Reviewers [azbreakchangereview@microsoft.com](mailto:azbreakchangereview@microsoft.com) and ARM can review the changes afterwards. Once you have the approval, please remove the "ARMChangesRequested" label from the PR to make the PR visible to the ARM reviewers.

mkatkam commented:
> > Hi @mkatkam! The automation detected breaking changes in this pull request. As a result, it added the `BreakingChangeReviewRequired` label. You cannot proceed with merging this PR until you complete one of the following action items: **ACTION ITEM ALTERNATIVE A**: Fix the breaking change. Please consult the documentation provided in the relevant validation failures. **ACTION ITEM ALTERNATIVE B**: Request approval. Alternatively, if you cannot fix the breaking changes, then you can request an approval for them. Please follow the process described in the [High-level Breaking Change Process doc](https://aka.ms/breakingchangesprocess#high-level-breaking-change-process). **ACTION ITEM ALTERNATIVE C**: Report false positive. If you think there are no breaking changes, i.e. the validation should pass yet it fails, then proceed as explained in **ACTION ITEM ALTERNATIVE B**. This applies even if the breaking change tool fails with internal runtime error. In such case a manual breaking change review is necessary.
> 
> Please fix the breaking changes or please get a breaking changes approval from Azure Breaking Changes Reviewers [azbreakchangereview@microsoft.com](mailto:azbreakchangereview@microsoft.com) and ARM can review the changes afterwards. Once you have the approval, please remove the "ARMChangesRequested" label from the PR to make the PR visible to the ARM reviewers.

Hi, The breaking changes tag was added because I changed the description of a property, but currently I removed everything and kept as is. So, currently there are no breaking changes.

TimLovellSmith commented:
>Hi, The breaking changes tag was added because I changed the description of a property,

But that doesn't really explain these errors does it?

>Rule	Message
❌ 1005 - RemovedPath	The new version is missing a path that was found in the old version. Was path '/{resourceUri}/providers/Microsoft.ResourceHealth/childAvailabilityStatuses/current' removed or restructured?
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L884:5
❌ 1005 - RemovedPath	The new version is missing a path that was found in the old version. Was path '/{resourceUri}/providers/Microsoft.ResourceHealth/childAvailabilityStatuses' removed or restructured?
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L931:5
❌ 1005 - RemovedPath	The new version is missing a path that was found in the old version. Was path '/{resourceUri}/providers/Microsoft.ResourceHealth/childResources' removed or restructured?
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L981:5
❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was 'articleId' renamed or removed?
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1563:11
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1710:11
❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was '_ActionUrl.Comment' renamed or removed?
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1748:7
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1899:7
❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was 'reason' renamed or removed?
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L987:11
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1134:11
❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was 'articleId' renamed or removed?
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1073:15
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1224:15
❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was 'parameters' renamed or removed?
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1073:15
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1224:15
⚠️ 1032 - DifferentAllOf	The new version has a different 'allOf' property than the previous one.
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1936:11
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L2091:11
⚠️ 1032 - DifferentAllOf	The new version has a different 'allOf' property than the previous one.
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L938:13
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1085:13
⚠️ 1032 - DifferentAllOf	The new version has a different 'allOf' property than the previous one.
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1946:5
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L2101:5
💬 1041 - AddedPropertyInResponse	The new version has a new property 'maintenanceId' in response that was not found in the old version.
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L987:11
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1134:11
💬 1041 - AddedPropertyInResponse	The new version has a new property 'maintenanceType' in response that was not found in the old version.
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L987:11
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1134:11
💬 1041 - AddedPropertyInResponse	The new version has a new property 'argQuery' in response that was not found in the old version.
New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L987:11
Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1134:11

TimLovellSmith commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+            },
+            "maintenanceId": {
+              "type": "string",
+              "description": "Unique Id for Planned maintenance event"
```
>Id [](http://example.com/codeflow?start=37&length=2)

identifier?

mkatkam replied:
I will fix this. Thank you!

TimLovellSmith commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+            },
+            "maintenanceId": {
+              "type": "string",
+              "description": "Unique Id for Planned maintenance event"
```
Doc style - they should be "planned maintenance events" (all lowercase) - no need to uppercase this in English docs.

mkatkam replied:
I will fix this, thank you!

TimLovellSmith commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+            },
+            "argQuery": {
+              "type": "string",
+              "description": "ARG Query to fetch the affected resources from their existing ARG locations"
```
You should expand out the acronym ARG in case its not obvious to people in the context of Microsoft.ResourceHealth docs

mkatkam replied:
I will expand ARG, thank you!

TimLovellSmith commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+        },
+        "extensionName": {
+          "type": "string",
+          "description": "It provides the name of portal extension to produce link for given service health event."
```
I suggest a doc style/content change "It provides the name of portal extension to produce link for given service health event." to just "The name of the Azure portal extension to link to for the given service health event."

mkatkam replied:
Quick question on this - We want to correct this and make it better, but will it not create another breaking change as I am changing something that is already existing.?

TimLovellSmith commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+        },
+        "bladeName": {
+          "type": "string",
+          "description": "It provides the name of portal extension blade to produce link for given service health event."
```
I suggest a doc style/content change to just "The name of the Azure portal extension blade to link to for the given service health event."

mkatkam replied:
Since this is already existing, I am wondering if this will be a breaking change if I change this.

TimLovellSmith commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+          "type": "string",
+          "description": "It provides the name of portal extension blade to produce link for given service health event."
+        },
+        "parameters": {
```
this object definition doesn't include a proper type declaration that explains how the mapping would be used by an API consumer.

TimLovellSmith commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json:
```diff
@@ -0,0 +1,2167 @@
+        },
+        "parameters": {
+          "type": "object",
+          "description": "It provides a map of parameter name and value for portal extension blade to produce lik for given service health event."
```
typo 'lik' - I wonder if its not caught by spellcheck. Should it be link?

mkatkam replied:
I will fix this,  but will check if this will cause any breaking changes if I change this. And then proceed accordingly.

TimLovellSmith commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/preview/2023-07-01-preview/examples/Events_ListBySubscriptionId.json:
```diff
@@ -0,0 +1,110 @@
+              "hirStage": "resolved",
+              "maintenanceId": "{maintenanceId}",
+              "maintenanceType": "{maintenanceType}",
+              "argQuery": "{argQuery}"
```
are these realistic enough values that you would want to show in your example docs?

mkatkam replied:
These are not realistic values, will try to add sample values for the examples. Thank you for your suggestion.

TimLovellSmith commented:
> > Hi, The breaking changes tag was added because I changed the description of a property,
> 
> But that doesn't really explain these errors does it?
> 
> > Rule	Message
> > ❌ 1005 - RemovedPath	The new version is missing a path that was found in the old version. Was path '/{resourceUri}/providers/Microsoft.ResourceHealth/childAvailabilityStatuses/current' removed or restructured?
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L884:5

From looking at the PR diff some more, looks like you based it off an old preview version 2022-10-01-preview instead of the latest stable version?

mkatkam commented:
> > > Hi, The breaking changes tag was added because I changed the description of a property,
> > 
> > 
> > But that doesn't really explain these errors does it?
> > > Rule	Message
> > > ❌ 1005 - RemovedPath	The new version is missing a path that was found in the old version. Was path '/{resourceUri}/providers/Microsoft.ResourceHealth/childAvailabilityStatuses/current' removed or restructured?
> > > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L884:5
> 
> From looking at the PR diff some more, looks like you based it off an old preview version 2022-10-01-preview instead of the latest stable version? If that's just accidental probably you should fix it, so people can you all your features with the new preview.

Yes, I confirmed with my team as well that I should be using the latest stable version as the base version. So, I will create another PR with base version as 2022-10-01. Is there any other way I can use the same PR ?

mkatkam commented:
> > Hi, The breaking changes tag was added because I changed the description of a property,
> 
> But that doesn't really explain these errors does it?
> 
> > Rule	Message
> > ❌ 1005 - RemovedPath	The new version is missing a path that was found in the old version. Was path '/{resourceUri}/providers/Microsoft.ResourceHealth/childAvailabilityStatuses/current' removed or restructured?
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L884:5
> > ❌ 1005 - RemovedPath	The new version is missing a path that was found in the old version. Was path '/{resourceUri}/providers/Microsoft.ResourceHealth/childAvailabilityStatuses' removed or restructured?
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L931:5
> > ❌ 1005 - RemovedPath	The new version is missing a path that was found in the old version. Was path '/{resourceUri}/providers/Microsoft.ResourceHealth/childResources' removed or restructured?
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L981:5
> > ❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was 'articleId' renamed or removed?
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1563:11
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1710:11
> > ❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was '_ActionUrl.Comment' renamed or removed?
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1748:7
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1899:7
> > ❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was 'reason' renamed or removed?
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L987:11
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1134:11
> > ❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was 'articleId' renamed or removed?
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1073:15
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1224:15
> > ❌ 1033 - RemovedProperty	The new version is missing a property found in the old version. Was 'parameters' renamed or removed?
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1073:15
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1224:15
> > ⚠️ 1032 - DifferentAllOf	The new version has a different 'allOf' property than the previous one.
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1936:11
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L2091:11
> > ⚠️ 1032 - DifferentAllOf	The new version has a different 'allOf' property than the previous one.
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L938:13
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1085:13
> > ⚠️ 1032 - DifferentAllOf	The new version has a different 'allOf' property than the previous one.
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L1946:5
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L2101:5
> > 💬 1041 - AddedPropertyInResponse	The new version has a new property 'maintenanceId' in response that was not found in the old version.
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L987:11
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1134:11
> > 💬 1041 - AddedPropertyInResponse	The new version has a new property 'maintenanceType' in response that was not found in the old version.
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L987:11
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1134:11
> > 💬 1041 - AddedPropertyInResponse	The new version has a new property 'argQuery' in response that was not found in the old version.
> > New: Microsoft.ResourceHealth/preview/2023-07-01-preview/ResourceHealth.json#L987:11
> > Old: Microsoft.ResourceHealth/stable/2022-10-01/ResourceHealth.json#L1134:11

Yes, I think not using the latest stable version as the base version caused this. I will fix these by using the latest stable version.

mkatkam commented:
Closing this PR as this is incorrectly based off of 2022-10-01-preview instead of 2022-10-01.
Original and new PR https://github.com/Azure/azure-rest-api-specs/pull/25447
